### ISSUE TITLE
Add Spark to the Docker image

### ISSUE THREAD

**Issue by mboes** [tweag/sparkle#45]
The Docker image contributed by @abailly does not include Spark. It's fine for `inline-java` users but for sparkle users it would be nice if this Docker image ultimately supported a workflow for non-Linux users (OS X, Windows) where they could test an app locally using `spark-submit`.

The issue is that installing Spark on Ubuntu isn't entirely trivial, and the `--extra-{include,library}-dirs` we already need when using Docker is a bit of a nuisance. So here's an idea: let's unify our Nix containerization and the Docker support by turning Nix into a provisioner for the Docker image. In that way, we can install all system dependencies using the same shared `shell.nix` file that `stack --nix build` uses.

In summary, we'd have

```
$ stack --nix build
```

as the recommended build command on Linux. And

```
$ stack --docker --nix build
```

for anyone not on Linux or anyone who doesn't want to install Nix on their host.

This ought to work, but there's apparently a bug upstream:

```
$ cat > Dockerfile <<EOF
FROM nixos/nix
RUN nix-env -i stack
EOF
$ docker build -t sparkle .
$ stack --docker --docker-image sparkle --docker-stack-exe image --no-docker-set-user --nix build
```

seems to ignore the `--nix` flag. @YPares any ideas?

**Comment by YPares** [comment]
@m There's indeed a flag (getReExec) in stack right now that prevents a nix-shell from being run inside a docker container. Deactivating it allows stack to attempt to run the nix-shell.

The slight issue with that approch I can think of is the fact that if you install spark via nix (or any dependency via nix) inside the container, then all the dependencies will be removed when the container stops running.
So you can use:
```
docker:
    persist: true
```
in your `stack.yaml`, but either:
- everytime you reboot your machine you'll have to re-install every; or
- you'll have to work off a custom image configured to contain your dependencies.

**Comment by YPares** [comment]
The last blocking step seem to be nix accessing the nixpkgs installed in the docker container.
Whatever image I'm based on (I tried sparkle's Dockerfile, nixos/nix & fpco/stack) when creating my docker image with eveything I need, I'm getting 
`error: file ‘nixpkgs’ was not found in the Nix search path (add it using $NIX_PATH or -I), at (string):1:14`
when running `stack --docker --nix --docker-image img build`
However. When I `docker run -ti img` inside the container I can perfectly run `nix-shell -p something` or `nix-env -i something`. It's the explicit mention of `<nixpkgs>` (like in the nix expression generated by stack) that makes it fail.

**Comment by YPares** [comment]
@m I put the current state of the work in https://github.com/tweag/sparkle/blob/nix-docker-config/docker-image.nix

The image is buildable and stack is able to relaunch itself in it (with `--docker-stack-exe host`!)
but sparkle build fails because zlib headers and .so cannot be found

**Comment by robinbb** [comment]
Is this issue fixed, now, with your recent merge, @mboes ?

**Comment by mboes** [comment]
Indeed.
